Merged OSGiTest and JavaOSGiTest into one Java file

tanyatanyatanya opened the pull request:
@htreu As you suggested I merged the OSGiTest and JavaOSGiTest files into one

maggu2810 commented:
> @htreu As you suggested I merged the OSGiTest and JavaOSGiTest files into one

IMHO we should get rid of the old Groovy based class and then perhaps rename the pure Java one.
Can you point me to that suggestion?

You introduce new package imports for Groovy in some manifests.
IMHO we should get rid of Groovy some time and not introduce new dependencies to that packages.

tanyatanyatanya commented:
@maggu2810 I removed the imports for Groovy in the manifests and if I understand you clearly, you suggest only to rename the JavaOSGiTest file?Also I deleted the OSGiTest Groovy file.

maggu2810 commented:
@tanyatanyatanya I don't know what @htreu has suggested.

IMHO there should be (some time) only one base OSGi test file and that is written in pure Java without any dependencies to Groovy.
Additional we should remove the usage of Groovy from all tests (it is a long way but we already state that we don't accept new Groovy contributions), so we should not add any new package imports in Manifests for Groovy packages.

tanyatanyatanya commented:
@maggu2810 @htreu Could you check the updated PR?

maggu2810 commented on bundles/test/org.eclipse.smarthome.test/src/main/java/org/eclipse/smarthome/test/java/JavaOSGiTest.java:
```diff
@@ -53,19 +54,18 @@
      * @return bundle context
      */
-    private @Nullable BundleContext initBundleContext() {
+    protected @Nullable BundleContext getBundleContext() {
```
The private method `initBundleContext` is used to initialize the member variable `bundleContext`.

The member variable is protected, so it can be accessed by subclasses.
It is also annotated so we ignore the nullness annotations and the fact that it could be null.

If you want to access the bundleContext member by a function, please introduce a new one `getBundleContext()` that returns the member.

There is no need to use `FrameworkUtil` all the time to access the bundle context (it should be enough to do it once in `@Before`).

htreu commented:
Imho we should in the end only have the `org.eclipse.smarthome.test.OSGiTest` since the `java` package and prefix was only there to distinguish between the groovy and java file.

htreu commented:
For legacy reasons it might be a good idea to keep the `org.eclipse.smarthome.test.java.JavaOSGiTest` and just let it extend from `OSGiTest` without any additional content. @tanyatanyatanya in case we agree on this, please add a comment in the `JavaOSGiTest` class that is is not intended for extension but the base `OSGiTest` should be used.
wdyt?

maggu2810 commented:
> Imho we should in the end only have the `org.eclipse.smarthome.test.OSGiTest` since the `java` package and prefix was only there to distinguish between the groovy and java file.

I agree.
I don't think we need to keep the empty `JavaOSGiTest` class. If someone use that class the code could be changed to use `OSGiTest`.

tanyatanyatanya commented:
@maggu2810 @htreu JavaOSGiTest usually extends JavaTest class and if I understand you correctly I should remove JavaTest and instead the extended class would be OSGiTest and JavaOSGiTest would be without any content just extension?

htreu commented:
No, here is the pan:
JavaTest will stay as it is, a base test class for none-OSGi-tests.
JavaOSGiTest and OSGiTest should now basically be the same, we simply want to drop the JavaOSGiTest.
I also do agree with @maggu2810 that we should not carry a empty version of it around.

In the end there will be two base test classes:
* JavaTest
* OSGiTest

htreu commented:
@tanyatanyatanya let me know if you have further questions on the migration.

htreu commented:
@tanyatanyatanya friendly ping: do you plan to finish this one any time soon? There is a fix for `waitForAssert` in this PR which should be merged asap. Thanks.

htreu commented:
wrt groovy dependencies: since the resulting OSGiTest is further used in groovy based tests and must be able to handle `groovy.lang.Closure` the dependency will stay.

tanyatanyatanya commented:
@htreu I have a question about some of the java OSGi classes that extend JavaOSGiTest. After the merge they will extend OSGiTest but everyone of them must import groovy.lang.Closure and if I understand correctly from your previous comment it is ok? Or should I make them extend the JavaTest one after including methods like getBundleContext and getService, so they can be pure java ones.

htreu commented:
Hi @tanyatanyatanya thanks for the feedback. I actually tried the merge myself just this morning and experienced the same issue. Unfortunately this is not what we want, my comment was only related to the smarthome.test bundle. All other plain java test fragments should not have an extra dependency to groovy.lang.

Long story short: I had an offline discussion with @sjka, @kaikreuzer and @triller-telekom where we agreed that we postpone the merge until all groovy tests are migrated. This means we will close this PR and see the `OSGiTest` from the groovy source folder as a legacy base test which will not be developed further. The `JavaOSGiTest` will stay as it is but improvements will only go there.

One last point: There is a bug fix for the OSGiTest#waitForAssert where the given Closure finally gets executed. @tanyatanyatanya please create a separate PR with this fix only. Thanks.

@maggu2810 any objections to this "plan"?

maggu2810 commented:
No objections, IMHO this fits to what I have written at the beginning.

tanyatanyatanya commented:
@htreu Just to be clear I should create now a new PR for: 
>a bug fix for the OSGiTest#waitForAssert where the given Closure finally gets executed.

Thanks.

tanyatanyatanya commented:
@htreu Just to double check, you mean this exact row - https://github.com/eclipse/smarthome/blob/f9cd85b14245b07cc5b18628ee7c03a3e91549b5/bundles/test/org.eclipse.smarthome.test/src/main/groovy/org/eclipse/smarthome/test/OSGiTest.groovy#L252 where previously I removed the assertion? Thanks.

htreu commented:
Yes I think so. It leads to the issue that when the timeout runs out the assertion is not finally executed and it will never fail. So a timeout will lead to an accepted test case (which is false most of the time).
